### DISCUSSION TITLE
Cross Scala Versions

### DISCUSSION THREAD

**Discussion by rlavolee** [hnaderi/edomata#155] [Ideas]
Hi @hnaderi !

I find your work pleasant to read and I would like to give it a try. Unfortunately I am still stuck with Scala 2.13. 
Are you interested in cross-publishing Edomata to Scala 3 and 2.13 ?

Regards,
Rémi

**Comment by hnaderi** [comment]
Hi @rlavolee :wave: 
Thank you!
While most of the modules are completely fine with cross publishing, I think there are also some meta programming stuff which is probably a considerable amount of work to back-port to scala 2.
I'll come back to you on this when I have checked the scope of required changes.
Also, in the meanwhile you can use `scala3from213` compile option (but I'm not sure about its friendliness or problems it might have) or try the edomata in a separate project.

**Reply by rlavolee** [reply]
Thanks. I will look into `scala3from213` compile option.

**Reply by rlavolee** [reply]
I'm sorry to bother you but I can't find informations related to `scala3from213` compile option. Can you point me resources about it ?

**Reply by hnaderi** [reply]
https://www.scala-lang.org/blog/2021/04/08/scala-3-in-sbt.html

**Reply by hnaderi** [reply]
tldr version:
```scala
// Use a Scala 3 library in Scala 2.13 
libraryDependency += ("org.bar" %% "bar" % "1.0.0").cross(CrossVersion.for2_13Use3)
```

**Reply by rlavolee** [reply]
Thanks 🙏

**Reply by rlavolee** [reply]
It actually resolved dependancy with `CrossVersion.for2_13Use3` option.
However I tried to replicate the bank account example but stumbled on error while implementing account decision.


```
Unsupported Scala 3 extension method with unmergeable type parameters: decide; found in trait edomata.core.ModelSyntax.

[error]   def open: Decision[Rejection, Event, Open] = decide { // 1
[error]                                                        ^
[error] one error found

```

`Rejection`, `Event` and `Open` are all sealed traits. I guess Tasty have limitations with Scala3 new features in Scala2.13.

**Reply by hnaderi** [reply]
I think you might have deleted some stuff in the editing process, `decide` is a syntax method and is available on domain models, however the code you have shared uses this function directly. (the example from site uses `this.decide`)

**Reply by hnaderi** [reply]
By the way, if you want to run the examples you can also use the already available example [here](https://github.com/hnaderi/edomata-example).

**Reply by rlavolee** [reply]
> By the way, if you want to run the examples you can also use the already available example [here](https://github.com/hnaderi/edomata-example).

That's a good idea. I forked the example and pushed a [`test-2.13`](https://github.com/rlavolee/edomata-example/tree/test-2.13) branch where you will find the isolated errors if you don't mind to take a look. 

> I think you might have deleted some stuff in the editing process, `decide` is a syntax method and is available on domain models, however the code you have shared uses this function directly. (the example from site uses `this.decide`)

I indeed removed `this.` because compiler was complaining that `decide` was not a member of Account.

```
value decide is not a member of dev.hnaderi.example.accounts.Account
```

**Reply by hnaderi** [reply]
I checked the branch you mentioned, the problem seems to be related to compile can't understand newer language features and can't be solved.

I also did some experiments to add scala 2.13 support, but it requires breaking changes in the APIs which some of them also degrade the user experience, as edomata tries to optimize user experience and performance using some of the scala3 exclusive features, such as inline methods and opaque types which are hard to backport to scala 2.
Given the fact that new scala projects are mostly using scala3 (according to my experience which isn't perfect), the effort to backport this features seems to be mostly waste.

May I ask you what is the use case you want to use edomata with 2.13? If there are use cases which I'm missing or anything I can help with, I'd be more than happy to help or add support for.

**Reply by rlavolee** [reply]
Thank you for having a look. I surely understand your points and I absolutly agree with. 

For the context, I'm currently using Play which is almost but not yet compatible with Scala3.

**Reply by g-dumas** [reply]
Thank you @hnaderi. We will use Edomata for sure in our next scala3 projects. 
Thank's a lot for your work.
